src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContains.cs

```diff
@@ -0,0 +1,197 @@
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!TryGetRequiredMethods(context.Compilation, out var containsMethod, out var addMethod, out var removeMethod))
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(context => OnConditional(context, containsMethod, addMethod, removeMethod), OperationKind.Conditional);
+        }
+
+        private static void OnConditional(OperationAnalysisContext context, IMethodSymbol containsMethod, IMethodSymbol addMethod, IMethodSymbol removeMethod)
+        {
+            var conditional = (IConditionalOperation)context.Operation;
+
+            if (!TryExtractContainsInvocation(conditional.Condition, containsMethod, out var containsInvocation, out var containsNegated))
+            {
+                return;
+            }
+
+            if (!TryExtractAddOrRemoveInvocation(conditional.WhenTrue.Children, addMethod, removeMethod, containsNegated, out var addOrRemoveInvocation))
+            {
+                return;
+            }
+
+            if (!AreInvocationsOnSameInstance(containsInvocation, addOrRemoveInvocation))
+            {
+                return;
+            }
+
+            using var locations = ArrayBuilder<Location>.GetInstance(2);
+            locations.Add(conditional.Syntax.GetLocation());
+            locations.Add(addOrRemoveInvocation.Syntax.Parent!.GetLocation());
+
+            context.ReportDiagnostic(containsInvocation.CreateDiagnostic(Rule, additionalLocations: locations.ToImmutable(), null));
+        }
+
+        private static bool TryGetRequiredMethods(
+            Compilation compilation,
+            [NotNullWhen(true)] out IMethodSymbol? containsMethod,
+            [NotNullWhen(true)] out IMethodSymbol? addMethod,
+            [NotNullWhen(true)] out IMethodSymbol? removeMethod)
+        {
+            var iSetType = WellKnownTypeProvider.GetOrCreate(compilation).GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericISet1);
+            var iCollectionType = WellKnownTypeProvider.GetOrCreate(compilation).GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericICollection1);
+
+            if (iSetType is null || iCollectionType is null)
+            {
+                containsMethod = null;
+                addMethod = null;
+                removeMethod = null;
+
+                return false;
+            }
+
+            addMethod = iSetType.GetMembers(Add).OfType<IMethodSymbol>().FirstOrDefault();
+            containsMethod = iCollectionType.GetMembers(Contains).OfType<IMethodSymbol>().FirstOrDefault();
+            removeMethod = iCollectionType.GetMembers(Remove).OfType<IMethodSymbol>().FirstOrDefault();
```

Comment:
Could use just one loop
```suggestion
            foreach (var method in iCollectionType.GetMembers().OfType<IMethodSymbol>())
            {
                switch (method.Name)
                {
                    case Remove: removeMethod = method; break;
                    case Contains: containsMethod = method; break;
                }
            }
```

Reply:
The analyzer needs both `Remove` and `Contains` from `ICollection<T>` -- this would only find one of them and then break out.

Reply:
No, the break is for switch case not for the loop

Reply:
Ohh, I see. I missed that, sorry. I will also apply the same change for `iImmutableSetType` and its methods.